### PR TITLE
DSND-2747: Improve performance of get list query

### PR DESCRIPTION
**Pull request by crowlandsCH**
## Describe the changes
* Get list  3 separate repository calls similar to this PR by Chris here https://github.com/companieshouse/company-appointments.api.ch.gov.uk/pull/169
* Finds a list of filing history document _id’s dependent on items per page, start index and category (if present) sorted descending on data.date. 
* Retrieves a list of FH documents to be present in the GET response using the list of id’s (preserving the sort order using a __sort_order__ field constructed on the fly to index the items in the filing history document list). 
* A final query to calculate the total_count of documents in the database matching the request sent.
* Various import and sonarlint issues fixed.

Tested locally, GET requests still return fine and reflect the database.

### Related Jira tickets

[DSND-2747](https://companieshouse.atlassian.net/browse/DSND-2747)

### General

- [x] Is the PR as small as it can be?
- [x] Has the Jira ticket been updated with any relevant comments?
- [ ] Is the `README` up to date?
- [ ] Has the `POM` been updated for the latest versions of dependencies?
- [ ] Has the code been double-checked against `main`?
- [ ] Does the code adhere standards in this repository, including SonarQube checks?

### Testing

- [ ] Do the code changes have unit and integration tests with code coverage > 80%?
- [ ] Has the code been tested locally and/or passed the API Karate tests on Tilt?
- [ ] Have mandatory manual test cases been run?
- [ ] Are extra Karate tests required?
- [ ] Are all the test data resources up to date?

### Release preparation

_Where possible, add links to the respective Jira tickets when an item is checked_

- [ ] Have these changes been included in a release ticket?
- [ ] Are changes required to the `docker-chs-development` deployment or test data?
- [ ] Are any changes required to the environment added to deployment repositories?

## Notes to the tester

_Add any testing hints or instructions here._

[DSND-2747]: https://companieshouse.atlassian.net/browse/DSND-2747?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/filinghistory/api/repository/Repository.java [hunk 1]
```diff
@@ -41,37 +40,63 @@ public Repository(MongoTemplate mongoTemplate) {
...
             Aggregation aggregation = newAggregation(
                     match(criteria),
-                    facet(
-                            count().as("count")).as("total_count")
-                            .and(match(new Criteria()), sort(Direction.DESC, "data.date")).as("document_list"),
-                    unwind("$total_count", true),
-                    project()
-                            .and(ifNull("$total_count.count").then(0)).as("total_count")
-                            .andExpression("$document_list").slice(itemsPerPage, startIndex).as("document_list"));
+                    facet(match(new Criteria()), sort(Direction.DESC, "data.date")).as("ids"),
```
Comment (cgregoryCH) [inline src/main/java/uk/gov/companieshouse/filinghistory/api/repository/Repository.java:52]: the match stage can be removed

File: src/test/java/uk/gov/companieshouse/filinghistory/api/repository/RepositoryTest.java [hunk 2]
```diff
@@ -44,42 +44,62 @@ class RepositoryTest {
...
         // given
         when(mongoTemplate.aggregate(any(), eq(FilingHistoryDocument.class),
-                eq(FilingHistoryListAggregate.class))).thenReturn(aggregationResults);
-        when(aggregationResults.getUniqueMappedResult()).thenReturn(listAggregate);
+                eq(FilingHistoryIds.class))).thenReturn(aggregationResultsFilingHistoryIds);
+        when(mongoTemplate.aggregate(any(), eq(FilingHistoryDocument.class),
+                eq(FilingHistoryDocument.class))).thenReturn(aggregationResultsFullDocuments);
+        when(aggregationResultsFilingHistoryIds.getUniqueMappedResult()).thenReturn(filingHistoryIds);
+        when(aggregationResultsFullDocuments.getMappedResults()).thenReturn(mockFilingHistoryDocuments);
+        when(mongoTemplate.count(any(), eq(FilingHistoryDocument.class))).thenReturn(Long.valueOf(1));
```
Comment (cgregoryCH) [inline src/test/java/uk/gov/companieshouse/filinghistory/api/repository/RepositoryTest.java:68]: ```Long.valueOf(1)``` can be replaced with ```1L```

File: src/test/java/uk/gov/companieshouse/filinghistory/api/repository/RepositoryTest.java [hunk 3]
```diff
@@ -44,42 +44,62 @@ class RepositoryTest {
...
```
Comment (cgregoryCH) [inline src/test/java/uk/gov/companieshouse/filinghistory/api/repository/RepositoryTest.java:91]: Same here

File: src/test/java/uk/gov/companieshouse/filinghistory/api/service/FilingHistoryServiceTest.java [hunk 4]
```diff
@@ -86,52 +89,53 @@ void findExistingFilingHistoryDocumentShouldReturnEmptyWhenNoDocumentExists() {
...
-    void findCompanyFilingHistoryListShouldCallRepositoryWithCorrectCategories(List<String> actualCategories,
+    void findCompanyFilingHistoryListQueriesShouldCallRepositoryWithCorrectCategories(List<String> actualCategories,
             List<String> expectedCategories) {
         // given
-        when(repository.findCompanyFilingHistory(any(), anyInt(), anyInt(), any()))
-                .thenReturn(filingHistoryListAggregate);
-        when(filingHistoryListAggregate.getTotalCount()).thenReturn(1);
+        when(repository.findListOfFilingHistoryIds(any(), anyInt(), anyInt(), any())).thenReturn(new FilingHistoryIds());
+        when(repository.findFullFilingHistoryDocuments(any())).thenReturn(new ArrayList<>());
+        when(repository.countTotal(any(), any())).thenReturn(Long.valueOf(1));
```
Comment (cgregoryCH) [inline src/test/java/uk/gov/companieshouse/filinghistory/api/service/FilingHistoryServiceTest.java:97]: Same here

File: src/test/java/uk/gov/companieshouse/filinghistory/api/service/FilingHistoryServiceTest.java [hunk 5]
```diff
@@ -86,52 +89,53 @@ void findExistingFilingHistoryDocumentShouldReturnEmptyWhenNoDocumentExists() {
...
     @Test
-    void findCompanyFilingHistoryListShouldCallRepositoryAndReturnEmptyWhenTotalCountZero() {
+    void findCompanyFilingHistoryListQueriesShouldCallRepositoryAndReturnEmptyWhenTotalCountZero() {
         // given
-        when(repository.findCompanyFilingHistory(any(), anyInt(), anyInt(), any()))
-                .thenReturn(filingHistoryListAggregate);
-        when(filingHistoryListAggregate.getTotalCount()).thenReturn(0);
+        when(repository.findListOfFilingHistoryIds(any(), anyInt(), anyInt(), any())).thenReturn(new FilingHistoryIds());
+        when(repository.findFullFilingHistoryDocuments(any())).thenReturn(new ArrayList<>());
+        when(repository.countTotal(any(), any())).thenReturn(Long.valueOf(0));
```
Comment (cgregoryCH) [inline src/test/java/uk/gov/companieshouse/filinghistory/api/service/FilingHistoryServiceTest.java:116]: Same here

File: src/test/java/uk/gov/companieshouse/filinghistory/api/repository/RepositoryTest.java [hunk 6]
```diff
@@ -44,42 +44,62 @@ class RepositoryTest {
...
+    @Mock
+    private List<FilingHistoryDocument> mockFilingHistoryDocuments;
     @Mock
     private AggregationResults<FilingHistoryDeleteAggregate> deleteAggregationResults;
     @Mock
     private FilingHistoryDeleteAggregate deleteAggregate;
 
     @Test
-    void shouldCallMongoTemplateWithCompanyNumberCriteriaOnly() {
+    void shouldCallMongoTemplateSuccessfullyWithCompanyNumberCriteriaOnly() {
```
Comment (cgregoryCH) [inline src/test/java/uk/gov/companieshouse/filinghistory/api/repository/RepositoryTest.java:60]: These tests are testing too much. There should be one repository method called per test which tests a specific part of that method. Single responsibility should apply to source code and unit tests.

**Review by cgregoryCH — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/companieshouse/filinghistory/api/repository/Repository.java [hunk 7]
```diff
@@ -41,37 +40,63 @@ public Repository(MongoTemplate mongoTemplate) {
         this.mongoTemplate = mongoTemplate;
     }
 
-    public FilingHistoryListAggregate findCompanyFilingHistory(String companyNumber,
-            int startIndex, int itemsPerPage, List<String> categories) {
+    public FilingHistoryIds findListOfFilingHistoryIds(String companyNumber,
```
[the remark this answers is not in the document]
Comment (crowlandsCH) [reply src/main/java/uk/gov/companieshouse/filinghistory/api/repository/Repository.java:43]: done

File: src/main/java/uk/gov/companieshouse/filinghistory/api/repository/Repository.java (hunk 1 above)
Comment (crowlandsCH) [reply src/main/java/uk/gov/companieshouse/filinghistory/api/repository/Repository.java:52]: done

File: src/test/java/uk/gov/companieshouse/filinghistory/api/repository/RepositoryTest.java (hunk 2 above)
Comment (crowlandsCH) [reply src/test/java/uk/gov/companieshouse/filinghistory/api/repository/RepositoryTest.java:68]: done

File: src/test/java/uk/gov/companieshouse/filinghistory/api/service/FilingHistoryServiceTest.java (hunk 4 above)
Comment (crowlandsCH) [reply src/test/java/uk/gov/companieshouse/filinghistory/api/service/FilingHistoryServiceTest.java:97]: done

File: src/test/java/uk/gov/companieshouse/filinghistory/api/service/FilingHistoryServiceTest.java (hunk 5 above)
Comment (crowlandsCH) [reply src/test/java/uk/gov/companieshouse/filinghistory/api/service/FilingHistoryServiceTest.java:116]: done

File: src/test/java/uk/gov/companieshouse/filinghistory/api/repository/RepositoryTest.java (hunk 3 above)
Comment (crowlandsCH) [reply src/test/java/uk/gov/companieshouse/filinghistory/api/repository/RepositoryTest.java:91]: done

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by cgregoryCH — APPROVED** [review]

**Review by johncook-ch — APPROVED** [review]

**Pull request merged**